bravado/mapping/unmarshal.py

```diff
@@ -0,0 +1,143 @@
+
+    else:
+        raise SwaggerError(
+            "Don't know how to unmarshal value {0} with a value of {1}"
+            .format(value, obj_type))
+
+
+def unmarshal_primitive(spec, value):
+    """Unmarshal a jsonschema primitive type into a python primitive.
+
+    :type spec: dict or jsonref.JsonRef
+    :type value: int, long, float, boolean, string, unicode, etc
+    :rtype: int, long, float, boolean, string, unicode, or an object
+        based on 'format'
+    :raises: TypeError
+    """
+    if value is None and schema.is_required(spec):
+        # TODO: Error message needs more context. Consider adding a stack like
+        #       `context` object to each `unmarshal_*` method that acts like
+        #       breadcrumbs.
+        raise TypeError('Spec {0} says this is a required value'.format(spec))
+
+    value = formatter.from_wire(spec, value)
+    jsonschema.validate(value, spec)
+    return value
+
+
+def unmarshal_array(swagger_spec, array_spec, array_value):
+    """Unmarshal a jsonschema type of 'array' into a python list.
+
+    :type swagger_spec: :class:`bravado.mapping.spec.Spec`
+    :type array_spec: dict or jsonref.JsonRef
+    :type array_value: list
+    :rtype: list
+    :raises: TypeError
+    """
+    if not is_list_like(array_value):
+        raise TypeError('Expected list like type for {0}:{1}'.format(
+            type(array_value), array_value))
+
+    # TODO: could also do this in-place instead of allocating a new array. Think
+    #       about implications of this some more...
+    result = []
+    for element in array_value:
+        result.append(unmarshal_schema_object(
+            swagger_spec, array_spec['items'], element))
+    return result
+
+
+def unmarshal_object(swagger_spec, object_spec, object_value):
+    """Unmarshal a jsonschema type of 'object' into a python dict.
+
+    :type swagger_spec: :class:`bravado.mapping.spec.Spec`
+    :type object_spec: dict or jsonref.JsonRef
+    :type object_value: dict
+    :rtype: dict
+    :raises: TypeError
+    """
+    if not is_dict_like(object_value):
+        raise TypeError('Expected dict like type for {0}:{1}'.format(
+            type(object_value), object_value))
+
+    # TODO: could also do this in-place instead of allocating a new dict. Think
+    #       about implications of this some more...
+    result = {}
+    props_spec = object_spec['properties']
+    for prop_name, prop_spec in props_spec.iteritems():
+        result[prop_name] = unmarshal_schema_object(
+            swagger_spec, prop_spec, object_value.get(prop_name))
+    return result
+
+
+def unmarshal_model(swagger_spec, model_spec, model_value):
+    """Unmarshal a dict into a Model instance.
+
+    :type swagger_spec: :class:`bravado.mapping.spec.Spec`
+    :type model_spec: dict or jsonref.JsonRef
+    :type model_value: dict
+    :rtype: Model instance
+    :raises: TypeError
+    """
+    model_name = model_spec[MODEL_MARKER]
```

Comment:
I believe `x-` are use extensions in swagger 2.0 , I'm actually not sure how that relates to `x-model` here.

Is this something that we're going to require for swagger-py support?